### PR TITLE
Don't apply element based offsets to text nodes in $patchStyleText

### LINKED ISSUE #3125: Bug: $patchStyleText only patches part of the children when the selection ends at the end of an element
Lexical version: 0.5.0

## Steps To Reproduce

1. Set up a selection whose right-hand side is the end of an _element_ (i.e. offset set to the number of children in the element)
2. Pass said selection to `$patchStyleText`

Link to code example: (PR forthcoming)

## The current behavior

Inside `$patchStyleText` we'll fetch the nodes in the selection using `RangeSelection.getNodes()`. This will return a series of nodes including the end element and its text children.

_However_ the last node returned will not be the element the selection ends on but rather its text children. As a result `lastNode` will point to a text node but `endOffset` refers to the child offset within the end element, _not_ an offset within `lastNode`.

Nevertheless, we'll go ahead and split the node using `endOffset`: https://github.com/facebook/lexical/blob/737f8811254ff499755ac5ab5535a736fe9b142a/packages/lexical-selection/src/index.ts#L429-L431

As a result, only a portion of the selected element's children will be formatted.

## The expected behavior

We shouldn't apply child-based offsets to the text of the text node children themselves.

Note that `insertText` has a similar bug but I haven't had a chance to file it yet.

### PR DESCRIPTION
**Pull request by birtles**
Fixes #3125

### CODE REVIEW

File: packages/lexical-selection/src/index.ts [hunk 1]
```diff
@@ -367,15 +367,14 @@ export function $patchStyleText(
   const firstNodeTextLength = firstNodeText.length;
   const focusOffset = focus.offset;
   let anchorOffset = anchor.offset;
-  let startOffset;
-  let endOffset;
   const isBefore = anchor.isBefore(focus);
-  startOffset = isBefore ? anchorOffset : focusOffset;
-  endOffset = isBefore ? focusOffset : anchorOffset;
+  const [start, end] = isBefore ? [anchor, focus] : [focus, anchor];
```
Comment (trueadm) [inline packages/lexical-selection/src/index.ts:371]: Do we need this change? It adds more logical runtime overhead compared to before.

File: packages/lexical-selection/src/index.ts [hunk 2]
```diff
@@ -425,6 +434,14 @@ export function $patchStyleText(
       const lastNodeText = lastNode.getTextContent();
       const lastNodeTextLength = lastNodeText.length;
 
+      // The last node might not actually be the end node
+      //
+      // If not, assume the last node is fully-selected unless the end offset is
+      // zero.
+      if (lastNode.getKey() !== end.key && endOffset) {
```
[the remark this answers is not in the document]
Comment (birtles) [reply packages/lexical-selection/src/index.ts]: Fixed.

File: packages/lexical-selection/src/index.ts (hunk 1 above)
Comment (birtles) [reply packages/lexical-selection/src/index.ts]: No, it's not needed. Removed.

**Comment by birtles** [comment]
I believe the test failures in e2e-mac are unrelated pre-existing flakiness.

**Comment by acywatson** [comment]
> I believe the test failures in e2e-mac are unrelated pre-existing flakiness.

Yea I think you're right - re-running tests now. If you can resolve conflicts here too, that'd be awesome. Sorry for the churn.

**Comment by birtles** [comment]
I'll rebase this once #3116 is merged since it will conflict with that.

**Comment by birtles** [comment]
Rebased and ready for review

**Review by acywatson — APPROVED** [review]

**Pull request merged**